packages/sdk/src/fee/FeeEstimator.ts

```diff
@@ -0,0 +1,164 @@
+import { Psbt } from "bitcoinjs-lib";
+import { Buffer } from "buffer";
+import { MAXIMUM_FEE } from "../constants";
+import { OrditSDKError } from "../errors";
+import { getNetwork, getScriptType } from "../utils";
+import type { AddressFormat } from "../addresses/types";
+import type { Network } from "../config/types";
+import type { FeeEstimatorOptions } from "./types";
+
+class FeeEstimator {
+  protected fee = 0;
+  protected feeRate: number;
+  protected network: Network;
+  protected psbt: Psbt;
+  protected witness?: Buffer[] = [];
+  protected virtualSize = 0;
+  protected weight = 0;
+
+  constructor({ feeRate, network, psbt, witness }: FeeEstimatorOptions) {
+    if (feeRate < 0 || !Number.isSafeInteger(feeRate)) {
+      throw new OrditSDKError("Invalid feeRate");
+    }
+
+    this.feeRate = +feeRate; // convert decimal to whole number that might have passed Number.isSafeInteger check due to precision loss
+    this.network = network;
+    this.witness = witness || [];
+    this.psbt = psbt || new Psbt({ network: getNetwork(this.network) });
+  }
+
+  get data() {
+    return {
+      fee: this.fee,
+      virtualSize: this.virtualSize,
+      weight: this.weight,
+    };
+  }
+
+  private sanityCheckFee() {
+    if (this.fee > MAXIMUM_FEE) {
+      throw new OrditSDKError("Error while calculating fees");
+    }
+  }
+
+  calculateNetworkFee(): number {
+    this.fee = this.calculateVirtualSize() * this.feeRate;
+    this.sanityCheckFee();
+
+    return this.fee;
+  }
+
+  private analyzePSBTComponents() {
+    const inputs = this.psbt.data.inputs;
+    const outputs = this.psbt.txOutputs;
+    const inputTypes: AddressFormat[] = [];
+    const outputTypes: AddressFormat[] = [];
+
+    if (inputs.length === 0) {
+      throw new OrditSDKError("PSBT must have at least one input");
+    }
+
+    if (outputs.length === 0) {
+      throw new OrditSDKError("PSBT must have at least one output");
+    }
+
+    inputs.forEach((input) => {
+      const script =
+        input.witnessUtxo && input.witnessUtxo.script
+          ? input.witnessUtxo.script
+          : null;
+
+      if (!script) {
+        throw new OrditSDKError("Invalid script");
+      }
+
+      inputTypes.push(getScriptType(script, this.network).format);
+    });
```

Comment:
@iamcrazycoder do the inputs for this function always include witness scripts? Is `legacy` (p2pkh) excluded since there is no witness script?

Reply:
Legacy doesn't have anything as witness script -- the reason there's a fallback to `null` on line 69.